Public/Src/Tools/SandboxedProcessExecutor/Executor.cs

```diff
@@ -0,0 +1,322 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using System.Threading.Tasks;
+using BuildXL.Native.IO;
+using BuildXL.Processes;
+using BuildXL.SandboxedProcessExecutor.Tracing;
+using BuildXL.Utilities;
+using BuildXL.Utilities.Instrumentation.Common;
+using BuildXL.Utilities.Tracing;
+
+namespace BuildXL.SandboxedProcessExecutor
+{
+    internal sealed class Executor
+    {
+        private const int ProcessRelauchCountMax = 5;
+
+        private readonly Configuration m_configuration;
+        private readonly LoggingContext m_loggingContext = new LoggingContext("BuildXL.SandboxedProcessExecutor");
+        public readonly TrackingEventListener TrackingEventListener = new TrackingEventListener(Events.Log);
+        private readonly Stopwatch m_telemetryStopwatch = new Stopwatch();
+        private OutputErrorObserver m_outputErrorObserver;
+
+        /// <summary>
+        /// Creates an instance of <see cref="Executor"/>.
+        /// </summary>
+        public Executor(Configuration configuration)
+        {
+            Contract.Requires(configuration != null);
+
+            m_configuration = configuration;
+    }
+
+        public int Run()
+        {
+            AppDomain.CurrentDomain.UnhandledException +=
+                (sender, eventArgs) =>
+                {
+                    HandleUnhandledFailure(eventArgs.ExceptionObject as Exception);
+                };
+
+            TelemetryStartup();
+
+            ExitCode exitCode = RunInternal();
+
+            TelemetryShutdown();
+
+            return (int)exitCode;
+        }
+
+        private void HandleUnhandledFailure(Exception exception)
+        {
+            // Show the exception to the user
+            ConsoleError(exception.ToString());
+            
+            // Log the exception to telemetry
+            if (AriaV2StaticState.IsEnabled)
+            {
+                Logger.Log.SandboxedProcessExecutorCatastrophicFailure(m_loggingContext, exception.ToString());
+                TelemetryShutdown();
+            }
+
+            Environment.Exit((int)ExitCode.InternalError);
+        }
+
+        private void TelemetryStartup()
+        {
+            if (!Debugger.IsAttached && m_configuration.EnableTelemetry)
+            {
+                AriaV2StaticState.Enable(global::BuildXL.Tracing.AriaTenantToken.Key);
+                TrackingEventListener.RegisterEventSource(ETWLogger.Log);
+                m_telemetryStopwatch.Start();
+            }
+        }
+
+        private void TelemetryShutdown()
+        {
+            if (AriaV2StaticState.IsEnabled && m_configuration.EnableTelemetry)
+            {
+                m_telemetryStopwatch.Stop();
+                Logger.Log.SandboxedProcessExecutorInvoked(m_loggingContext, m_telemetryStopwatch.ElapsedMilliseconds, Environment.CommandLine);

```

Comment:
>Logger.Log.SandboxedProcessExecutorInvoked(m_loggingContext, m_telemetryStopwatch.ElapsedMilliseconds, Environment.CommandLine); [](start = 16, length = 128)

I'm worried about this scaling especially considering the full command line is being sent to telemetry. There's also connection setup/teardown overhead by hooking the process into telemetry. Do you think we can get away with just logging crashes to telemetry and relying on the calling build engines to log the fact that the pip was executed?

Reply:
As discussed, I'll change the default enableTelemetry to false, and make it false if it runs under a build engine.

---
In reply to: [276484358](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276484358) [](ancestors = 276484358)